skada/_subspace.py

```diff
@@ -1021,3 +1025,341 @@ def TransferSubspaceLearning(
+            J_ct = (
+                (1 / n_s**2) * Beta.t() @ K_tilde_s @ Beta
+                - (2 / n_s * n_t)
+                * torch.ones((n_t, 1), dtype=torch.float64).t()
+                @ K_tilde_t_s
+                @ Beta
+                + (1 / n_t**2)
+                * torch.ones((n_t, 1), dtype=torch.float64).t()
+                @ K_t
+                @ torch.ones((n_t, 1), dtype=torch.float64)
+            )
+
+            Jreg = (self.lmbd_s / n_s) * torch.norm(
+                A - torch.ones((n_components, n_s), dtype=torch.float64), p=2
+            ) + (self.lmbd_l / n_s) * torch.norm(B, p=2)
+
+            J_ct_con = (
+                J_ct
+                + self.lmbd
+                * self.eps
+                * torch.trace(
+                    L
+                    @ torch.inverse(
+                        K_tilde_s + n_s * self.eps * torch.eye(n_s, dtype=torch.float64)
+                    )
+                )
+                + Jreg
+            )
+
+            return J_ct_con
+
+        #####
+        X_source = torch.tensor(X_source, dtype=torch.float64)
+        X_target = torch.tensor(X_target, dtype=torch.float64)
+        y_source = torch.tensor(y_source, dtype=torch.float64)
+
+        alpha = torch.ones(n_c, dtype=torch.float64)
+        W = torch.ones((D, n_components), dtype=torch.float64)
+        G = torch.ones((n_c, n_components), dtype=torch.float64)
+        H = torch.zeros((n_c, n_components), dtype=torch.float64)
+
+        Beta, A, B, R_dis = compute_Beta_A_R(alpha, G, H)
+
+        for i in range(self.max_iter):
+            if i % 3 == 0:
+                # For α, we use quadratic programming (QP) to minimize
+                # Jˆct w.r.t. α under constraint (11)
+                # Define constraints
+                alpha_constraints = (
+                    {"type": "ineq", "fun": (lambda x: -(R_dis.detach().cpu() @ x))},
+                    {"type": "eq", "fun": (lambda x: np.sum(x) - 1)},
+                )
+
+                result = minimize(
+                    func_np,
+                    x0=alpha,
+                    args=(W, G, H),
+                    method="SLSQP",
+                    constraints=alpha_constraints,
+                    options={"maxiter": 1},
+                )
+
+                alpha = result.x
+                alpha = torch.tensor(alpha, dtype=torch.float64)
+            elif i % 3 == 1:
+                # Define constraint function for Grassmann manifold: W^TW = I_d
+                # def grassmann_constraint_function(W):
+                #     return np.dot(W.T, W) - np.eye(n_components)
+
+                # grassmann_constraint = {
+                #     "type": "eq",
+                #     "fun": grassmann_constraint_function,
+                # }
+
+                (_, W, _, _), _ = torch_minimize(
+                    func_torch, (alpha, W, G, H), tol=self.tol, max_iter=1
+                )
+                W = torch.tensor(W, dtype=torch.float64)
```

Comment:
Here any idea how to enforce the constraint to the torch minimiser ?

Reply:
hello, here are a couple of ideas:
- Use the dual of the optimization problem resulting in adding the constraint to the loss in Pytorch:
min_x  f(x) st g(x) = 0 <==> min_{x, lambda} f(x) + lambda * g(x)
here the constraint would be the norm of W^TW - Id

- Another idea could be to generate the matrices directly on the Grassmanian manifold:
 a) geomtorch: https://geotorch.readthedocs.io/en/latest/orthogonal/grassmannian.html 

b) torch: https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.orthogonal.html
Une autre methode, c'est de parametriser tes fonctions tq ce soit orthogonal avec 

- Maybe not directly related, but here is an interesting paper with code to optimize orthogonal matrices: 
https://proceedings.mlr.press/v151/ablin22a/ablin22a.pdf (code: https://github.com/pierreablin/landing)